SourceCode/include/tracer/Ray.h

```diff
@@ -8,29 +8,25 @@
 struct Intersection {
   Vector hitPoint;
   Vector hitNormal;
+
+// Ugly, I know :-D
+#if defined(BARYCENTRIC) && BARYCENTRIC
+  float u;
+  float v;
+#endif  // BARYCENTRIC
 };
 
 enum RayType { Primary, Shadow };
 
-template <RayType T = Primary>
 class Ray {
  public:
   Vector origin;
   Vector direction;
+  RayType rayType = Primary;
 
   Ray();
-  Ray(const Vector &origin, const Vector &direction);
-  // Ray(const Ray &other) = default;
-  // Ray &operator=(const Ray &other) = default;
-  // Ray(Ray &&other) = default;
-  // Ray &operator=(Ray &&other) = default;
+  Ray(const Vector &origin, const Vector &direction, const RayType &rayType = Primary);
 
   // TODO (maybe move to Triangle.h)
   std::optional<Intersection> intersectWithTriangle(const Triangle &triangle, const bool smoothShading = false) const;
```

Comment:
You have to be very careful with std in such delicate places (code executed a lot of times could be million of times)